cpp/src/arrow/util/decimal.cc

```diff
@@ -29,8 +29,8 @@ ARROW_EXPORT Status FromString(
   }
 
   int8_t sign = 1;
-  auto charp = s.cbegin();
-  auto end = s.cend();
+  auto charp = s.c_str();
+  auto end = charp + s.length();
```

Comment:
@rip-nsk making comments like "parser is not good anyway" is not very consistent with the Apache Way and could read as disrespectful. 

We all want to make great software here, and we value your contributions. See https://www.apache.org/foundation/how-it-works.html#philosophy for more about what kind of conduct is expected of ASF contributors.